classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,218 @@
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
+        // From stdClass to merge_request class.
+        $mergerequest = merge_request::from($mergerequest);
+        if ($mergerequest->retries > $maxattempts) {
+            $log_attempts = array("Reached the number of maximum attempts.");
+            $mergerequest->append_log($log_attempts, time());
+            $this->update_status_and_log($mergerequest->id,
+                                        merge_request::COMPLETED_WITH_ERRORS,
+                                        $mergerequest->log);
+            return;
+        }
+        $this->verify_users_to_keep_and_remove($mergerequest);
+        $mergerequestresult = $this->merge($mergerequest,
+                                            $maxattempts,
+                                            merge_request::TRIED_WITH_ERROR);
+        if ($mergerequestresult->status == merge_request::COMPLETED_WITH_SUCCESS) {
+            /* We run the merge request AGAIN because the user may be interacting with Moodle
+            * while merge request is being processed, so that NO ALL records are correctly migrated
+            * into the user to keep. It will ensure ALL records are correctly migrated into the user to keep. */
+            $mergerequestresult = $this->merge($mergerequestresult,
+                                                $maxattempts,
+                                                merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(merge_request $mergerequest,
+                            int $maxattempts,
+                            int $statusforerror) {
+        $retries = $mergerequest->retries + 1;
+        $this->update_retries($mergerequest);
+        $mut = new MergeUserTool();
+        list($success, $log) = $mut->merge($mergerequest->keepuserid,
+                                            $mergerequest->removeuserid);
+        if (!$success) {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        } else {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        }
+        $mergerequest->append_log($log, time());
+        if ($mergerequest->retries > $maxattempts) {
+            $log_attempts = "Reached the number of maximum attempts";
```

Comment:
the same here. maybe it can be a method within the `merge_request` class, like `check_if_maxattempts_was_reached()` and do all this `if` inside.

Reply:
lines like `$log_attempts = get_string('maxattemptsreached', 'tool_mergeusers');`

should be like:

`$log_attempts = get_string('maxattemptsreached', 'tool_mergeusers', (object)['retries'=>$mergerequest->retries, 'maxretries' => $maxattempts]);`

in order to report both numbers in the resulting string. Update the lang file properly too.

I commented this in some other comment, but I cannot find it over here.